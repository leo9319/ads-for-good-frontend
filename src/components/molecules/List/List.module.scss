@import '../../../styles/colours';
@import '../../../styles/spacing';

@mixin titleText($color, $fontSize, $fontWeight, $lineHeight) {
  color: $color;
  font-weight: $fontWeight;
  font-size: var($fontSize);
  line-height: var($lineHeight);
}

.list {
  background-color: transparent;
  width: 100%;
  display: flex;
  padding-bottom: var(--spacing-lg);
  flex-direction: column;
  align-items: flex-start;
  gap: var(--spacing-md);
  align-self: stretch;

  &_inner {
    width: 100%;
    gap: 0px;
  }
}

.listItem,
.addMoreButtonBox {
  padding: var(--spacing-sm) 0;
  align-self: stretch;
  border-bottom: 1px solid $color-neutral-300;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: var(--spacing-sm);

  .start {
    .avatar {
      display: flex;
      width: 48px;
      height: 48px;
      justify-content: center;
      align-items: center;
      border-radius: 999px;
    }

    .startInner {
      display: flex;
      flex-direction: column;
      align-items: flex-start;
      gap: var(--spacing-2xs);
      align-self: stretch;

      .title {
        @include titleText(
          $color-neutral-950,
          --typography-font-size-2,
          600,
          --typography-line-height-2
        );
      }

      .subtitle {
        @include titleText(
          $color-neutral-600,
          --typography-font-size-1,
          400,
          --typography-line-height-1
        );
      }
    }

    .addButtonCaption {
      @include titleText(
        $color-blue-500,
        --typography-font-size-1,
        700,
        --typography-line-height-1
      );
    }
  }

  .end {
    @include titleText(
      $color-neutral-950,
      --typography-font-size-2,
      600,
      --typography-line-height-2
    );
  }

  &_Flex {
    gap: var(--spacing-sm);
    flex: 1 0 0;
  }

  &.sizeLarge {
    .start .startInner .title,
    .end {
      @include titleText(
        $color-neutral-950,
        --typography-font-size-3,
        600,
        --typography-line-height-3
      );
    }

    .start .startInner .subtitle {
      @include titleText(
        $color-neutral-600,
        --typography-font-size-2,
        400,
        --typography-line-height-2
      );
    }
  }
}

.listItem:first-child {
  border-top: 1px solid $color-neutral-300;
}

.loadMoreBox {
  display: flex;
  padding: var(--spacing-2xl) 0px;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: var(--spacing-xs, 8px);
  align-self: stretch;

  .loadMore {
    width: 100%;
    font-size: var(--typography-font-size-3) !important;

    @media (min-width: 1280px) {
      // Note: We don't have 20px in spacing variables,
      // so need to confirm it with design team
      padding: var(--spacing-sm) 20px !important;
    }
  }
}

.tagBox {
  * {
    background-color: transparent !important;
  }
}

.listModal {
  .listItem,
  .addMoreButtonBox {
    border-bottom: 1px solid $color-neutral-100;
  }

  .listItem:first-child {
    border-top: 1px solid $color-neutral-100;
  }
}
