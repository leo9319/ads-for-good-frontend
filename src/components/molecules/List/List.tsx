import React from 'react';
import { DataList } from '@radix-ui/themes';
import Image from '@components/atoms/Image';
import Icon from '@components/atoms/Icons';
import Box from '@radix-styles/atoms/Box';
import Button from '@radix-styles/atoms/Button';
import Flex from '@radix-styles/atoms/Flex';
import Tag from '@radix-styles/atoms/Tag';
import Text from '@radix-styles/atoms/Text';
import { classNames } from '@utils/common/classNames';

import FallbackImage from '@assets/images/FallbackImage.svg';
import styles from './List.module.scss';

/**
 * Dynamic object representing each list item.
 * Values can be string, number, boolean, null, undefined, or a ReactNode.
 * Keys are flexible based on provided keyMapper.
 */
export interface listItemProps {
  [key: string]: string | number | boolean | null | undefined | React.ReactNode;
}
export interface ListButtonProps {
  /**
   * Text to be displayed on the button.
   * @property
   */
  text: string;
  /**
   * Show loading spinner if true.
   * @property
   */
  loading?: boolean;

  /**
   * Disable the button if true.
   * @property
   */
  disabled?: boolean;

  /**
   * Controls visibility of the button.
   * @property
   */
  isVisible?: boolean;
  /**
   * Callback function to handle button click.
   * @property
   */
  onClick?: () => void;
}

export interface ListProps {
  /**
   * Layout style of the list: default or modal.
   */
  listType?: 'default' | 'modal';

  /**
   * Data array to populate the list items.
   */
  data: listItemProps[];

  /**
   * Configuration for the "Load More" button.
   */
  loadMore?: ListButtonProps;

  /**
   * Configuration for the "Add More" inline item.
   */
  addMore?: ListButtonProps;

  /**
   * Type of content to render on the right side of each list item.
   * Can be a plain string, image, or custom React node.
   */
  endContent?: 'string' | 'image' | 'child';

  /**
   * Maps keys from the data object to be used in rendering list items.
   */
  keyMapper: {
    /**
     * Key name in `data` used for the main title.
     */
    title: string;

    /**
     * Key name in `data` used for the subtitle.
     */
    subTitle: string;

    /**
     * Key name in `data` used for rendering the right-side content.
     */
    endContent?: string;

    /**
     * Key name in `data` pointing to the image URL used in the `Tag` component.
     */
    tagURL?: string;

    /**
     * Key name in `data` providing the label text for the `Tag`.
     */
    tagLabel?: string;

    /**
     * Key name in `data` providing the avatar image URL shown at the start.
     */
    avatar?: string;
  };
  size: 'small' | 'large';
}
/**
 * List Component
 *
 * Renders a flexible list UI using Radix UI's `DataList`.
 * It supports dynamic key mapping for data rendering, optional avatars, tag elements,
 * and configurable end content (text, image, or custom element).
 * Also includes optional "Load More" and "Add More" controls.
 *
 * @component
 *
 * @param {ListProps} props - Props for the List component
 * @param {'default' | 'modal'} [props.listType='default'] - Determines list layout style.
 * @param {listItemProps[]} props.data - Array of objects representing each list row.
 * @param {'string' | 'image' | 'child'} [props.endContent='string'] - Type of end content to render on each row.
 * @param {Object} props.keyMapper - Key mapper to extract fields from each data object.
 *
 * @returns {JSX.Element} Rendered List component.
 *
 * @example
 * <List
 *   listType="modal"
 *   data={[{ name: 'John', age: 30 }]}
 *   keyMapper={{
 *     title: 'name',
 *     subTitle: 'age',
 *   }}
 * />
 */
export const List = ({
  listType = 'default',
  data,
  loadMore,
  addMore,
  endContent = 'string',
  keyMapper,
  size = 'small',
}: ListProps) => {
  const listClass = classNames(
    styles.list,
    listType === 'modal' ? styles.listModal : ''
  );
  const listItemClass = classNames(
    styles.listItem,
    size === 'large' ? styles.sizeLarge : ''
  );

  const handleClick = () => {
    if (loadMore && loadMore.onClick) loadMore.onClick();
  };
  return (
    <Box className={listClass}>
      <DataList.Root className={styles.list_inner}>
        {data.map((list: listItemProps, index: number) => (
          <DataList.Item align="center" key={index} className={listItemClass}>
            <Flex
              display="flex"
              gap="11px"
              align="center"
              className={styles.start}
            >
              {keyMapper?.avatar && list[keyMapper?.avatar] && (
                <Image
                  src={list[keyMapper?.avatar] as string}
                  alt="Globe Hero Image"
                  fallbackSrc={FallbackImage}
                  className={styles.avatar}
                />
              )}
              <Box className={styles.startInner}>
                <Text as="p" className={styles.title}>
                  {list[keyMapper.title] ?? ''}
                </Text>
                <Text as="p" className={styles.subtitle}>
                  {list[keyMapper.subTitle] ?? ''}
                </Text>
                {keyMapper?.tagURL &&
                  keyMapper.tagLabel &&
                  list[keyMapper?.tagURL] &&
                  list[keyMapper.tagLabel] && (
                    <Box className={styles.tagBox}>
                      <Tag
                        iconUrl={list[keyMapper.tagURL] as string}
                        mode="default"
                        showAvatar
                        text={list[keyMapper.tagLabel] as string}
                      />
                    </Box>
                  )}
              </Box>
            </Flex>
            {keyMapper.endContent && list[keyMapper.endContent] && (
              <Box>
                {endContent === 'string' && (
                  <Text as="p" className={styles.end}>
                    ${list[keyMapper.endContent] ?? ''}
                  </Text>
                )}
                {endContent === 'image' && (
                  <Image
                    src={list[keyMapper.endContent] as string}
                    alt="Globe Hero Image"
                    fallbackSrc={FallbackImage}
                    className={styles.globeImage}
                  />
                )}
                {endContent === 'child' && list[keyMapper.endContent]}
              </Box>
            )}
          </DataList.Item>
        ))}
        {addMore && addMore.isVisible && (
          <DataList.Item align="center" className={styles.addMoreButtonBox}>
            <Flex
              display="flex"
              gap="11px"
              align="center"
              className={styles.start}
            >
              <Icon
                backgroundColor="#F4F5F7"
                backgroundPadding="12px"
                color="#65676B"
                name="icon-plus"
                rounded
                size="24"
              />
              <Text as="p" className={styles.addButtonCaption}>
                {addMore.text}
              </Text>
            </Flex>
          </DataList.Item>
        )}
      </DataList.Root>
      {loadMore && loadMore.isVisible && (
        <Box className={styles.loadMoreBox}>
          <Button
            size="lg"
            text={loadMore.text}
            mode="outline"
            disabled={loadMore.disabled}
            loading={loadMore.loading}
            className={styles.loadMore}
            onClick={handleClick}
          />
        </Box>
      )}
    </Box>
  );
};

export default List;
