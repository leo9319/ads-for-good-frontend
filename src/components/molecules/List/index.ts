export { default } from './List';
export { List } from './List';
export type { listItemProps, ListButtonProps, ListProps } from './List';
