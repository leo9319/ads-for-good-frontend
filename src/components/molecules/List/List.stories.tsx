import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import Theme from '@radix-styles/atoms/Theme';
import List from './List';

import ListImage from '../../../assets/images/ListImage.svg';

const meta = {
  title: 'Library/Molecules/List',
  component: List,
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    data: {
      control: {
        type: 'object',
      },
    },
    keyMapper: {
      control: {
        type: 'object',
      },
    },
    size: {
      control: 'select',
      options: ['small', 'large'],
    },
    listType: {
      control: 'select',
      options: ['default', 'modal'],
    },
  },
} satisfies Meta<typeof List>;

export default meta;

type Story = StoryObj<typeof meta>;
export const DonationHistory: Story = {
  args: {
    data: [
      {
        name: 'Raw Hope monthly pledge',
        amount: 59.0,
        date: 'Dec 28, 2023',
      },
      {
        name: 'Water Appeal Christmas Donation',
        amount: 49.0,
        date: 'Nov 28, 2023',
      },
      {
        name: 'Monthly gift',
        amount: 59.0,
        date: 'Aug 28, 2023',
      },
      {
        name: 'One time gift',
        amount: 120.0,
        date: 'Aug 28, 2023',
      },
      {
        name: 'Areas of Most Need',
        amount: 120.0,
        date: 'Aug 28, 2023',
      },
      {
        name: 'Israel/Gaza Fund',
        amount: 120.0,
        date: 'Aug 28, 2023',
      },
      {
        name: 'Childcare Pledge Support',
        amount: 49.0,
        date: 'Aug 28, 2023',
      },
      {
        name: 'Monthly gift',
        amount: 59.0,
        date: 'Aug 29, 2023',
      },
      {
        name: 'Monthly gift',
        amount: 59.0,
        date: 'Aug 28, 2023',
      },
      {
        name: 'Monthly gift',
        amount: 59.0,
        date: 'Aug 28, 2023',
      },
    ],
    keyMapper: {
      title: 'name',
      subTitle: 'date',
      endContent: 'amount',
    },
    size: 'small',
    loadMore: {
      text: 'Load More',
      isVisible: true,
      onClick: () => {
        console.log('Load More');
      },
    },
  },
};
export const SponsorChild: Story = {
  args: {
    data: [
      {
        name: 'Raw Hope monthly pledge',
        url: ListImage,
        date: 'Dec 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Water Appeal Christmas Donation',
        url: ListImage,
        date: 'Nov 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Monthly gift',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'One time gift',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Areas of Most Need',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Israel/Gaza Fund',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Childcare Pledge Support',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Monthly gift',
        url: ListImage,
        date: 'Aug 29, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Monthly gift',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
      {
        name: 'Monthly gift',
        url: ListImage,
        date: 'Aug 28, 2023',
        tagUrl:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
        tagLabel: 'Letter to Ashraful',
      },
    ],
    keyMapper: {
      title: 'name',
      subTitle: 'date',
      endContent: 'url',
      tagURL: 'tagUrl',
      tagLabel: 'tagLabel',
    },
    size: 'small',
    endContent: 'image',
    loadMore: {
      text: 'Load More',
      isVisible: true,
      onClick: () => {
        console.log('Load More');
      },
    },
  },
};

export const ChildDetail: Story = {
  args: {
    size: 'small',
    data: [
      {
        title: 'Family',
        value: 'Lives with parents. 2 brothers and one sister.',
      },
      {
        title: 'Birthday',
        value: 'Nov 14, 2013',
      },
      {
        title: 'Education',
        value: 'Grade 4',
      },
      {
        title: 'Interests',
        value: 'Cricket',
      },
      {
        title: 'Current health',
        value: 'In good health.',
      },
      {
        title: 'Label',
        value: 'Content',
      },
    ],
    keyMapper: {
      title: 'title',
      subTitle: 'value',
    },
    listType: 'modal',
  },
};

export const ChildrenList: Story = {
  args: {
    size: 'small',
    listType: 'modal',
    data: [
      {
        title: 'Ashraful Siam',
        value: '10 years old · Ukhiya, Bangladesh',
        avatar:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
      },
      {
        title: 'Sofia Dariani',
        value: '4 years old · Quito, Equador',
        avatar:
          'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
      },
    ],
    keyMapper: {
      title: 'title',
      subTitle: 'value',
      avatar: 'avatar',
    },
    addMore: {
      isVisible: true,
      text: 'Sponsor Another Child',
      onClick: () => {
        console.log('Sponsor Another Child');
      },
    },
  },
};
