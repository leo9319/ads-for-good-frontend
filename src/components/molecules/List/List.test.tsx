import React from 'react';
import { render } from '@testing-library/react';
import List, { ListProps } from './List';

const mockData = [
  {
    name: 'Raw Hope monthly pledge',
    amount: 59.0,
    date: 'Dec 28, 2023',
  },
  {
    name: 'Water Appeal Christmas Donation',
    amount: 49.0,
    date: 'Nov 28, 2023',
  },
  {
    name: 'Monthly gift',
    amount: 59.0,
    date: 'Aug 28, 2023',
  },
  {
    name: 'One time gift',
    amount: 120.0,
    date: 'Aug 28, 2023',
  },
];
const keyMapper: ListProps['keyMapper'] = {
  title: 'name',
  subTitle: 'date',
  endContent: 'amount',
};
describe('List Component', () => {
  it('renders list items based on data', () => {
    const { baseElement } = render(
      <List size="small" data={mockData} keyMapper={keyMapper} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
