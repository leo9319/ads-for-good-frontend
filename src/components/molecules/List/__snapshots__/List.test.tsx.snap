// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component renders list items based on data 1`] = `
<body>
  <div>
    <div
      class="rt-Box list"
    >
      <dl
        class="rt-Text rt-DataListRoot rt-r-orientation-horizontal rt-r-size-2 list_inner"
      >
        <div
          class="rt-DataListItem rt-r-ai-center listItem"
        >
          <div
            class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-gap start"
            style="--gap: 11px;"
          >
            <div
              class="rt-Box startInner"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title"
                id=""
              >
                Raw Hope monthly pledge
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subtitle"
                id=""
              >
                Dec 28, 2023
              </p>
            </div>
          </div>
          <div
            class="rt-Box end"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default"
            >
              $
              59
            </p>
          </div>
        </div>
        <div
          class="rt-DataListItem rt-r-ai-center listItem"
        >
          <div
            class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-gap start"
            style="--gap: 11px;"
          >
            <div
              class="rt-Box startInner"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title"
                id=""
              >
                Water Appeal Christmas Donation
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subtitle"
                id=""
              >
                Nov 28, 2023
              </p>
            </div>
          </div>
          <div
            class="rt-Box end"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default"
            >
              $
              49
            </p>
          </div>
        </div>
        <div
          class="rt-DataListItem rt-r-ai-center listItem"
        >
          <div
            class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-gap start"
            style="--gap: 11px;"
          >
            <div
              class="rt-Box startInner"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title"
                id=""
              >
                Monthly gift
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subtitle"
                id=""
              >
                Aug 28, 2023
              </p>
            </div>
          </div>
          <div
            class="rt-Box end"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default"
            >
              $
              59
            </p>
          </div>
        </div>
        <div
          class="rt-DataListItem rt-r-ai-center listItem"
        >
          <div
            class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-gap start"
            style="--gap: 11px;"
          >
            <div
              class="rt-Box startInner"
            >
              <p
                class="rt-Text mocked-class-name typography-font-family-default title"
                id=""
              >
                One time gift
              </p>
              <p
                class="rt-Text mocked-class-name typography-font-family-default subtitle"
                id=""
              >
                Aug 28, 2023
              </p>
            </div>
          </div>
          <div
            class="rt-Box end"
          >
            <p
              class="rt-Text mocked-class-name typography-font-family-default"
            >
              $
              120
            </p>
          </div>
        </div>
      </dl>
    </div>
  </div>
</body>
`;
